native/service/header_sync/neo3/utils.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2020 The poly network Authors
+ * This file is part of The poly network library.
+ *
+ * The  poly network  is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The  poly network  is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The poly network .  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package neo3
+
+import (
+	"fmt"
+	"github.com/joeqian10/neo3-gogogo/crypto"
+	"github.com/joeqian10/neo3-gogogo/sc"
+	"github.com/joeqian10/neo3-gogogo/tx"
+	"github.com/polynetwork/poly/common"
+	cstates "github.com/polynetwork/poly/core/states"
+	"github.com/polynetwork/poly/native"
+	"github.com/polynetwork/poly/native/service/governance/neo3_state_manager"
+	hscommon "github.com/polynetwork/poly/native/service/header_sync/common"
+	"github.com/polynetwork/poly/native/service/utils"
+)
+
+//verify header of any height
+//find key height and get neoconsensus first, then check the witness
+func verifyHeader(native *native.NativeService, chainID uint64, header *NeoBlockHeader) error {
+	neoConsensus, err := getConsensusValByChainId(native, chainID)
+	if err != nil {
+		return fmt.Errorf("verifyHeader, get Consensus error: %s", err)
+	}
+	if neoConsensus.NextConsensus != header.Witness.GetScriptHash() {
+		return fmt.Errorf("verifyHeader, invalid script hash in header error, expected: %s, got: %s", neoConsensus.NextConsensus.String(), header.Witness.GetScriptHash().String())
+	}
+
+	msg, err := header.GetMessage()
+	if err != nil {
+		return fmt.Errorf("verifyHeader, unable to get hash data of header")
+	}
+	if verified := tx.VerifyMultiSignatureWitness(msg, header.Witness); !verified {
+		return fmt.Errorf("verifyHeader, VerifyMultiSignatureWitness error: %s, height: %d", err, header.GetIndex())
+	}
+	return nil
+}
+
+func VerifyCrossChainMsgSig(native *native.NativeService, magic uint32, crossChainMsg *NeoCrossChainMsg) error {
+	// todo, review code, get neo3 state validator from native contract
+	svListBytes, err := neo3_state_manager.GetCurrentStateValidator(native)
+	if err != nil {
+		return fmt.Errorf("verifyCrossChainMsg, neo3_state_manager.GetCurrentStateValidator error: %v", err)
+	}
+	svStrings, err := neo3_state_manager.DeserializeStringArray(svListBytes)
```

Comment:
you serialize it in `GetCurrentStateValidator` function and deserialize it again, why?